File: tech_session/lib/tech_session/control_flow.exs
defmodule ControlFlow do
  # ...

  def handle_call({:position_island, player, key, row, col}, _from, state) do
    board = player_board(state, player)

    with {:ok, rules} <- Rules.check(state.rules, {:position_islands, player}),
         {:ok, coordinate} <- Coordinate.new(row, col),
         {:ok, island} <- Island.new(key, coordinate),
         %{} = board <- Board.position_island(board, key, island) do
      state
      |> update_board(player, board)
      |> update_rules(rules)
      |> reply_success(:ok)
    else
      :error -> reply_error(state, :error)
      {:error, :invalid_coordinate} -> reply_error(state, :invalid_coordinate)
      {:error, :invalid_island_type} -> reply_error(state, :invalid_island_type)
      {:error, :overlapping_island} -> reply_error(state, :overlapping_island)
    end
  end

  # ...
end

File: tech_session/lib/tech_session/control_flow.js
//
//
//
//
//
//
//
//

export function positionIsland({ player, key, row, col }, state) {
  const board = player_board(state, player);

  const rules = Rules.check(state.rules, { type: "position_islands", player });
  if (!rules) {
    return reply_error(state, "Error");
  }

  const coordinate = Coordinate.new(row, col);
  if (!coordinate) {
    return reply_error(state, "Invalid coordinate");
  }

  const island = Island.new(key, coordinate);
  if (!island) {
    return reply_error(state, "Invalid coordinate");
  }

  const board = Board.position_island(board, key, island);
  if (!island) {
    reply_error(state, "Overlapping island");
  }

  const newState = update_rules(update_board(state, player, board), rules);

  return reply_success(newState, "Ok");
}

//
//
//
//
//
//
//
//
//
//
//
//
//
//
//

export function positionIsland({ player, key, row, col }, state) {
  const board = player_board(state, player);

  try {
    const rules = Rules.check(state.rules, {
      type: "position_islands",
      player
    });
    const coordinate = Coordinate.new(row, col);
    const island = Island.new(key, coordinate);
    const board = Board.position_island(board, key, island);

    const newState = update_rules(update_board(state, player, board), rules);

    return reply_success(newState, "Ok");
  } catch (error) {
    return reply_error(state, error);
  }
}
